File: platforms/ios/Packages/FunputKit/Sources/KeyboardInput/PersonalSuggestions/AuthoredTokenTracker.swift
struct AuthoredTokenTracker {
    static let minimumLength = 2
    static let maximumLength = 32

    private(set) var prefix = ""
    private(set) var completedToken: String?
    /// The word the next one will be recorded as following, or nil when nothing
    /// here can vouch for what came before.
    ///
    /// This lives in the tracker because the tracker is the only thing that sees
    /// every mutation — which separator ended a word, when the caret left, when
    /// the prefix was abandoned. Anything downstream sees the result and not the
    /// reason.
    private(set) var context: String?
    private var prefixLength = 0
    private var overflowCount = 0

    mutating func recordInsertion(_ text: String) {
        if text.utf8.count == 1, let byte = text.utf8.first {
            if byte >= 65 && byte <= 90 || byte >= 97 && byte <= 122 {
                append(text)
            } else {
                complete(separator: Character(UnicodeScalar(byte)))
            }
            return
        }
        for character in text {
            if character.isLetter || isCombiningMark(character) && (!prefix.isEmpty || overflowCount > 0) {
                append(String(character))
            } else {
                complete(separator: character)
            }
        }
    }

    mutating func recordDeletion() {
        completedToken = nil
        if overflowCount > 0 {
            overflowCount -= 1
        } else if !prefix.isEmpty {
            prefix.removeLast()
            prefixLength -= 1
        } else {
            // Nothing left of this word to delete, so the caret has crossed back
            // over a boundary and what precedes it is no longer known.
            context = nil
        }
    }

    mutating func apply(_ mutations: [DocumentMutation]) {
        for mutation in mutations {
            switch mutation {
            case let .deleteBackward(count):
                for _ in 0..<count { recordDeletion() }
            case let .insert(text):
                recordInsertion(text)
            case .moveCursor:
                // The caret left the token being tracked; nothing typed here is a word any
                // more, so the prefix is abandoned rather than continued.
                reset()
            }
        }
    }

    mutating func consume() -> KeyboardSuggestionInputUpdate {
        let update = KeyboardSuggestionInputUpdate(
            prefix: overflowCount == 0 ? prefix : "",
            completedToken: completedToken,
            context: context
        )
        completedToken = nil
        return update
    }

    mutating func reset() {
        prefix.removeAll(keepingCapacity: true)
        completedToken = nil
        context = nil
        prefixLength = 0
        overflowCount = 0
    }

    private mutating func append(_ text: String) {
        completedToken = nil
        if prefixLength < Self.maximumLength, overflowCount == 0 {
            prefix.append(contentsOf: text)
            prefixLength += 1
        } else {
            overflowCount += 1
        }
    }

    /// A space continues the sentence, so the word just finished becomes the
    /// context for the next one. Anything else — a full stop, a newline, a comma
    /// — ends it, and two words either side of that were never adjacent. Guessing
    /// the other way would teach pairs nobody typed.
    private mutating func complete(separator: Character) {
        if overflowCount == 0, prefixLength >= Self.minimumLength {
            completedToken = prefix
        }
        context = separator == " " ? completedToken : nil
        prefix.removeAll(keepingCapacity: true)
        prefixLength = 0
        overflowCount = 0
    }

    private func isCombiningMark(_ character: Character) -> Bool {
        character.unicodeScalars.allSatisfy { scalar in
            switch scalar.properties.generalCategory {
            case .nonspacingMark, .spacingMark, .enclosingMark: true
            default: false
            }
        }
    }
}

public struct KeyboardSuggestionInputUpdate: Equatable, Sendable {
    public let prefix: String
    public let completedToken: String?
    /// The word `prefix` is being typed after, when one can be vouched for.
    public let context: String?

    public init(prefix: String, completedToken: String?, context: String? = nil) {
        self.prefix = prefix
        self.completedToken = completedToken
        self.context = context
    }

    public static let empty = KeyboardSuggestionInputUpdate(prefix: "", completedToken: nil)
}

File: platforms/android/ime/src/main/java/app/funput/funput/ime/editing/AuthoredTokenTracker.kt
package app.funput.funput.ime.editing

internal data class AuthoredSuggestionUpdate(
    val prefix: String,
    val completedToken: String?,
    /** The word [prefix] is being typed after, when one can be vouched for. */
    val context: String? = null,
) {
    companion object {
        val Empty = AuthoredSuggestionUpdate("", null)
    }
}

internal class AuthoredTokenTracker {
    private var prefix = ""
    private var completedToken: String? = null

    /**
     * The word the next one will be recorded as following, or null when nothing
     * here can vouch for what came before.
     *
     * It lives in the tracker because the tracker is the only thing that sees
     * every edit — which separator ended a word, when the caret left, when the
     * prefix was abandoned. Everything downstream sees the result, not the reason.
     */
    private var context: String? = null
    private var directOverflow = false

    fun update(composingText: String, completed: String?, completedOnSpace: Boolean) {
        directOverflow = false
        prefix = composingText.takeIf(::isValidToken).orEmpty()
        completedToken = completed?.takeIf { isValidToken(it) && scalarCount(it) >= MinimumLength }
        if (completed != null) context = completedToken.takeIf { completedOnSpace }
    }

    fun accepted(candidate: String) {
        directOverflow = false
        prefix = ""
        completedToken = candidate.takeIf(::isValidToken)
        // Accepting a candidate writes the word and a space after it.
        context = completedToken
    }

    fun input(text: String) {
        var index = 0
        while (index < text.length) {
            val codePoint = text.codePointAt(index)
            if (isTokenScalar(codePoint)) {
                if (!directOverflow && scalarCount(prefix) < MaximumLength) {
                    prefix += String(Character.toChars(codePoint))
                } else {
                    prefix = ""
                    completedToken = null
                    directOverflow = true
                }
            } else {
                completedToken = prefix.takeIf {
                    !directOverflow && scalarCount(it) >= MinimumLength
                }
                // A space continues the sentence, so the word just finished is
                // what the next one follows. Anything else ends it, and the two
                // words either side were never adjacent.
                context = completedToken.takeIf { codePoint == ' '.code }
                prefix = ""
                directOverflow = false
            }
            index += Character.charCount(codePoint)
        }
    }

    fun backspace() {
        if (directOverflow || prefix.isEmpty()) {
            // Nothing left of this word to delete, so the caret has crossed back
            // over a boundary and what precedes it is no longer known.
            context = null
            return
        }
        prefix = prefix.dropLast(Character.charCount(prefix.codePointBefore(prefix.length)))
        completedToken = null
    }

    fun consume(): AuthoredSuggestionUpdate =
        AuthoredSuggestionUpdate(prefix, completedToken, context).also { completedToken = null }

    fun currentPrefix(): String = prefix

    fun reset() {
        directOverflow = false
        prefix = ""
        completedToken = null
        context = null
    }

    private fun isValidToken(text: String): Boolean {
        if (text.isEmpty()) return false
        var index = 0
        var count = 0
        while (index < text.length && count <= MaximumLength) {
            val codePoint = text.codePointAt(index)
            if (!isTokenScalar(codePoint)) return false
            index += Character.charCount(codePoint)
            count += 1
        }
        return count in 1..MaximumLength
    }

    private fun scalarCount(text: String): Int = text.codePointCount(0, text.length)

    /**
     * Letters and their marks, and nothing else — a digit ends a word here as it
     * does on iOS.
     *
     * The two trackers have to agree, or the same typing teaches two different
     * lexicons. Letters is the rule that agrees with what is being learned:
     * Vietnamese syllables contain no digits, so a run of them is never a word,
     * and treating it as one spends a slot out of five thousand on a date or a
     * price — and writes a phone number into the store on disk.
     *
     * It only comes up with the number row turned on. Otherwise digits are typed
     * from the symbols panel, where nothing is learned at all.
     */
    private fun isTokenScalar(codePoint: Int) =
        Character.isLetter(codePoint) || isCombiningMark(codePoint)

    private fun isCombiningMark(codePoint: Int): Boolean = when (Character.getType(codePoint)) {
        Character.NON_SPACING_MARK.toInt(),
        Character.COMBINING_SPACING_MARK.toInt(),
        Character.ENCLOSING_MARK.toInt(),
        -> true
        else -> false
    }

    private companion object {
        const val MinimumLength = 2
        const val MaximumLength = 32
    }
}
